eosio.system/src/rex.cpp

```diff
@@ -0,0 +1,484 @@
+      if( bitr == _rexbalance.end() ) {
+         _rexbalance.emplace( from, [&]( auto& rb ) {
+            rb.owner       = from;
+            rb.vote_stake  = amount;
+            rb.rex_balance = rex_received;
+         });
+      } else {
+         _rexbalance.modify( bitr, 0, [&]( auto& rb ) {
+            rb.vote_stake.amount  += amount.amount;
+            rb.rex_balance.amount += rex_received.amount;
+         });
+      }
+
+      update_voting_power( from, amount );
+
+      runrex(2);
+   }
+
+   /**
+    * Converts REX stake back into SYS tokens at current exchange rate
+    */
+   void system_contract::sellrex( account_name from, asset rex ) {
+
+      runrex(2);
+
+      require_auth( from );
+
+      auto itr = _rextable.begin();
+      eosio_assert( itr != _rextable.end(), "rex system not initialized yet" );
+
+      auto bitr = _rexbalance.find( from );
+      eosio_assert( bitr != _rexbalance.end(), "user must first buyrex" );
+      eosio_assert( bitr->rex_balance.symbol == rex.symbol, "asset symbol must be (4, REX)" );
+      eosio_assert( bitr->rex_balance >= rex, "insufficient funds" );
+
+      auto result = close_rex_order( bitr, rex );
+      if( std::get<0>(result) ) {
+         /// sellrex has been processed successfuly, transfer tokens and update voting power
+         INLINE_ACTION_SENDER(eosio::token, transfer)( N(eosio.token), { N(eosio.rex), N(active) },
+                                                       { N(eosio.rex), from, asset( std::get<1>(result), system_token_symbol ), "sell REX" } );
+         update_voting_power( from, asset( -(std::get<2>(result)), system_token_symbol ) );
+      } else {
+         /**
+          * REX order couldn't be filled and is added to queue.
+          * If account already has an open order, requested rex is added to existing order.
+          * If account has a closed order, action fails and account must claimrex first in order
+          * to delete closed order from queue.
+          */
+         rex_order_table rex_orders( _self, _self );
+         auto oitr = rex_orders.find( from );
+         if( oitr == rex_orders.end() ) {
+            rex_orders.emplace( from, [&]( auto& ordr ) {
+               ordr.owner         = from;
+               ordr.rex_requested = rex;
+               ordr.order_time    = current_time_point();
+            });
+         } else {
+            eosio_assert( oitr->is_open, "user has a closed rex order in queue, must claimrex first");
+            rex_orders.modify( oitr, 0, [&]( auto& ordr ) {
+               ordr.rex_requested.amount += rex.amount;
+               eosio_assert( bitr->rex_balance >= ordr.rex_requested, "insufficient funds for current and scheduled orders");
+            });
+         }
+      }
+   }
+
+   void system_contract::cnclrexorder( account_name owner ) {
+
+      require_auth( owner );
+
+      rex_order_table rexorders( _self, _self );
+      auto itr = rexorders.find( owner );
```

Comment:
you could simplify this with something like:

auto& order = rexorder.get(owner, "no sellrex order is scheduled");

Reply:
Simplified using require_find().